doc/lib.rst

```diff
@@ -17,6 +17,11 @@ low-level interface to a C library.
 
 Read this `document <apis.html>`_ for a quick overview of the API design.
 
+Note that there exists a category of `std/` libraries that are not accessible by the usual `import <library name>` but only via `import std/<library name>` although all of the ordinary standard libraries are accessible by the former import statement format.  For this reason, it is recommended that all standard library names be imported as per the second format as that will include all standard library categories, both old and new, with the only exceptions when there are libraries with the same name both in the `std` category and not in it.  In this case, if one wants the non-`std/` version, the first form must be used.
```

Comment:
I don't think this section is necessary tbh. `import std/<module name>` should be the usual way to import a module in the first place, so we shouldn't advertise it as an "alternative" way. For this reason, the modules docs also show the fully qualified name (i.e. with `std/` prefix, even if they're not in the `std/` folder). What could stay is the explanation why the `std/` prefix is recommended/mandatory, but please try to split your paragraphs in multiple lines, as done in the rest of the document.

Also note the terminology: The standard library is the whole thing, the different modules are called modules, not standard libraries.

Reply:
Thank you for your constructive suggestions:

As to terminology, yes, I see your point, and my references to standard libraries should be to standard **modules**.

As to the need for the section, I added it because, without digging deeply into the manual, one doesn't find the recommended new standard way of importing using `import std/<module name>' and then it is [in the section about pseudo import paths](https://nim-lang.org/docs/manual.html#modules-pseudo-importslashinclude-paths).

It is perhaps unfortunate that the pseudo name for all standard libraries is `std/` but `std/` is also a sub directory of modules and there is no documentation explaining the types of module that are of this sub folder and that these modules can only be imported using the new `std/` way whereas all the modules in the other sub folders will be automatically imported using either the old non-`std/` way and the new way.

You are right that this section needs to be reworded and shortened to reflect exactly what the standard library user needs to know.  Perhaps that section should be shortened to just the following?
```
Note that there exists a category of `std/` modules that are not accessible by the usual `import <module name>` but only via import `std/<module name>` although all of the ordinary standard modules are accessible by the former import statement format. For this reason, it is recommended that all standard modules be imported as per the second format as that will include all standard module categories, both old and new.

Read more about importing standard modules [in the manual](https://nim-lang.org/docs/manual.html#modules-pseudo-importslashinclude-paths).
```

Further suggestions would be welcome.

Reply:
> `import <module_name>` seems to now be intended to include both and all search paths including `std/` and `pkg/` but it doesn't currently find the modules in the `std/` **folder** unless `std/` is specified.

And why would that be desired if one just wants to import a stdlib module?

> Perhaps if we knew what types of modules are placed in this folder, it would be obvious why they can only be imported as part of the `std/` pseudo category.

That shouldn't be necessary, since everyone should use the `std/` prefix anyway.

> """
> Note that there exists a category of `std/` modules that are not accessible by the general `import <module name>` but only via `import std/<module name>` although all of the other standard modules are accessible by the broad import statement format. For this reason, it is recommended that all standard modules be imported as per the second format as that will include all standard module categories.
> 
> Read more about importing standard modules [in the manual](https://nim-lang.org/docs/manual.html#modules-pseudo-importslashinclude-paths).
> """

This sounds like `import <module name>` was the recommended default way to import stdlib modules, but it's not. IMO it should just not be mentioned here. Also note that the module docs already show `std/<module name>`.

Reply:
> That shouldn't be necessary, since everyone should use the std/ prefix anyway.

My confusion and my perceived need for the section arises from my understanding of the manual on these "Pseudo import/include paths" which says "They **can** be used to avoid ambiguity when there are multiple modules with the same path." with the bolding mine.  The word **"can"** does not mean **"must"**.

The sentence implies that one **may** use the pseudo path to avoid ambiguity such as if a package module had the same name as a standard library module.  It also implies that if one **doesn't** use a pseudo directory, then all paths will be searched, including all of the standard library and all dependent packages.  It **doesn't say** that if the `std/` pseudo path is not used, all paths will be searched **except for** the actual `std/` folder modules.

It would make sense not to include the `std/` folder for general searches if the modules in that path are experimental and shouldn't be imported except deliberately, but I see no documentation anywhere that states that general rule.  Since no `std/` library module is accepted that has the same name as a non-`std/` path module (which is why `system.channels_builtin` and `std/channels` have different names), there is no ambiguity and an import without any pseudo modifiers should find `channels`, but it doesn't.  This is inconsistent, and why I have tried to document it.

If standard library modules should **only** be imported using `std/` and dependent package modules should **only** be imported with `pkg/`, then not using either should be an error, but of course that currently can't be as there are too many modules that import without using either which is legal because the manual for this section doesn't say it is required, only that it is an option to avoid ambiguity.

Do you see my point?

Reply:
Then the manual should be reworded, so that it's clear that the `std/` prefix **should** be used, to avoid possible name conflicts (and it may become mandatory in the future) and that not using it doesn't work with some stdlib modules.

Reply:
@konsumlamm:

Following @Araq's idea, I now suggest this section be simplified to the following:

"""
The preferred way of importing entire single standard library modules is `import std/<module_name>` and importing entire multiple standard library modules is `import std/[module_name1, module_name2, ...]`; this will guarantee that **only and all** of the standard library paths are searched.

Read more about [importing standard modules in the manual](https://nim-lang.org/docs/manual.html#modules-pseudo-importslashinclude-paths).
"""

That seems to direct both new and experienced Nim programmers to the recommended way of importing standard modules that will always work, and the link can be followed for further information from the manual.

Reply:
Sounds good, but I'd reformulate it a bit:

> The preferred way of importing a single standard library module is `import std/<module_name>` and for multiple modules `import std/[<module_name1>, <module_name2>, ...]`; this will guarantee that **only** the standard library paths are searched.
>
> Read more about [importing std modules in the manual](https://nim-lang.org/docs/manual.html#modules-pseudo-importslashinclude-paths).

Reply:
@konsumlamm :

You are suggesting I remove the few words that describe **exactly** what adding `std/` to the import does as in **only** searching the standard library modules but also including **all** the standard library modules in the search (as opposed to not including those modules in the `std/` folder)?

You also suggest removing the words **entire**, which I added to indicate that these forms can be used to import all publicly exported items in the modules as opposed to the forms using `from` and `except` that can be used to only get partially imported modules?

I only included these few words to give correctness to the description, expecting users to follow the link to the manual for more complete descriptions.

Reply:
> You are suggesting I remove the few words that describe **exactly** what adding `std/` to the import does as in **only** searching the standard library modules but also including **all** the standard library modules in the search (as opposed to not including those modules in the `std/` folder)?

That seems redundant to me, "only searching the stdlib modules" implies "including all the stdlib modules in the search", the modules in the `std/` folder are part of the stdlib modules.



> You also suggest removing the words **entire**, which I added to indicate that these forms can be used to import all publicly exported items in the modules as opposed to the forms using `from` and `except` that can be used to only get partially imported modules?

Well, the recommended way is in fact to import the whole module ("entire" sounds weird here imo), so I don't see the problem, readers should be smart enough to infer that the recommended way to import only specific items is `from std/<module name> import ...`.

Reply:
I'm bored with this; have it your way.  Will you make the changes or should I?

Reply:
Feel free to make the changes.